en/concept/architecture.md

```diff
@@ -1,36 +1,121 @@
+PX4 consists of two main layers: the [flight stack](#flight-stack) is an autopilot software solution,
+and the [middleware](#middleware) is a general robotics layer that can support any type of autonomous robot.
 
-All [airframes](../airframes/architecture.md), and in fact all robotic systems including boats, share a single codebase. The complete system design is [reactive](http://www.reactivemanifesto.org), which means that:
+All PX4 [airframes](../airframes/README.md) share a single codebase (this includes other robotic systems like boats, rovers, submarines etc.). The complete system design is [reactive](http://www.reactivemanifesto.org), which means that:
 
-  * All functionality is divided into exchangeable components
-  * Communication is done by asynchronous message passing 
-  * The system can deal with varying workload
+- All functionality is divided into exchangeable components
+- Communication is done by asynchronous message passing 
+- The system can deal with varying workload
 
 In addition to these runtime considerations, its modularity maximizes [reusability](https://en.wikipedia.org/wiki/Reusability).
 
+
 ## High Level Software Architecture
 
-Each of the blocks below is a separate module, which is self-contained in terms of code, dependencies and even at runtime. Each arrow is a connection through publish/subscribe calls through [uORB](../middleware/uorb.md).
+The diagram below provides an overview of the building blocks of PX4. 
+The top part of the diagram contains middleware blocks, while the lower
+section shows the components of the flight stack.
+
+![PX4 Architecture](../../assets/diagrams/PX4_Architecture.svg)
+
+<!-- This diagram can be updated from 
+[here](https://drive.google.com/file/d/0B1TDW9ajamYkaGx3R0xGb1NaeU0/view?usp=sharing) 
+and opened with draw.io Diagrams. -->
+
+The source code is split into self-contained modules (shown in `monospace` in the
+diagram). Usually a building block corresponds to exactly one module. At
+runtime, you can inspect which modules are executed with the `top` command, and
+each module can be started/stopped individually via `<module_name> start/stop`.
+
+The arrows show the information flow for the *most important* connections between
+the modules. In reality, there are many more connections than shown, and some messages 
+(e.g. for parameters) are accessed by most of the modules.
+
+Modules communicate with each other through a publish-subscribe message bus
+named [uORB](../middleware/uorb.md). 
+The use of the publish-subscribe scheme means that:
+
+- The system is reactive — it will update instantly when new data is available
+- All operations and communication are fully parallelized
+- A system component can consume data from anywhere in a thread-safe fashion
+
+> **Info** This architecture allows every single one of these
+> blocks to be rapidly and easily replaced, even at runtime.
+
+The controllers / mixers are specific to a particular airframe (e.g. a
```

Comment:
We don't actually have any components called "mixers" on the diagram, and we haven't defined what either a controller or a mixer is. It would be good to do so somewhere here.